Mathlib/Logic/Equiv/Option.lean

```diff
@@ -0,0 +1,264 @@
+    some (removeNone_aux e x) = e none := by
+  simp [removeNone_aux, Option.not_isSome_iff_eq_none.mpr h]
+#align equiv.remove_none_aux_none Equiv.removeNone_aux_none
+
+theorem removeNone_aux_inv (x : α) : removeNone_aux e.symm (removeNone_aux e x) = x :=
+  Option.some_injective _
+    (by
+      cases h1 : e.symm (some (removeNone_aux e x)) <;> cases h2 : e (some x)
+      · rw [removeNone_aux_none _ h1]
+        exact (e.eq_symm_apply.mpr h2).symm
+
+      · rw [removeNone_aux_some _ ⟨_, h2⟩] at h1
+        simp at h1
+
+      · rw [removeNone_aux_none _ h2] at h1
+        simp at h1
+
+      · rw [removeNone_aux_some _ ⟨_, h1⟩]
+        rw [removeNone_aux_some _ ⟨_, h2⟩]
+        simp
+        )
+#align equiv.remove_none_aux_inv Equiv.removeNone_aux_inv
+
+/-- Given an equivalence between two `Option` types, eliminate `none` from that equivalence by
+mapping `e.symm none` to `e none`. -/
+def removeNone : α ≃ β where
+  toFun := removeNone_aux e
+  invFun := removeNone_aux e.symm
+  left_inv := removeNone_aux_inv e
+  right_inv := removeNone_aux_inv e.symm
+#align equiv.remove_none Equiv.removeNone
+
+@[simp]
+theorem remove_none_symm : (removeNone e).symm = removeNone e.symm :=
+  rfl
+#align equiv.remove_none_symm Equiv.remove_none_symm
+
+theorem remove_none_some {x : α} (h : ∃ x', e (some x) = some x') :
+    some (removeNone e x) = e (some x) :=
+  removeNone_aux_some e h
+#align equiv.remove_none_some Equiv.remove_none_some
+
+theorem remove_none_none {x : α} (h : e (some x) = none) : some (removeNone e x) = e none :=
+  removeNone_aux_none e h
+#align equiv.remove_none_none Equiv.remove_none_none
+
+@[simp]
+theorem option_symm_apply_none_iff : e.symm none = none ↔ e none = none :=
+  ⟨fun h => by simpa using (congr_arg e h).symm, fun h => by simpa using (congr_arg e.symm h).symm⟩
+#align equiv.option_symm_apply_none_iff Equiv.option_symm_apply_none_iff
+
+theorem some_removeNone_iff {x : α} : some (removeNone e x) = e none ↔ e.symm none = some x := by
+  cases' h : e (some x) with a
+  · rw [remove_none_none _ h]
+    simpa using (congr_arg e.symm h).symm
+
+  · rw [remove_none_some _ ⟨a, h⟩]
+    have h1 := congr_arg e.symm h
+    rw [symm_apply_apply] at h1
+    simp only [false_iff_iff, apply_eq_iff_eq]
+    simp [h1, apply_eq_iff_eq]
+
+#align equiv.some_remove_none_iff Equiv.some_removeNone_iff
+
+@[simp]
+theorem removeNone_option_congr (e : α ≃ β) : removeNone e.optionCongr = e :=
+  Equiv.ext fun x => Option.some_injective _ <| remove_none_some _ ⟨e x, by simp [EquivFunctor.map]⟩
+#align equiv.remove_none_option_congr Equiv.removeNone_option_congr
+
+end RemoveNone
+
+theorem optionCongr_injective : Function.Injective (optionCongr : α ≃ β → Option α ≃ Option β) :=
+  Function.LeftInverse.injective removeNone_option_congr
+#align equiv.option_congr_injective Equiv.optionCongr_injective
+
+set_option maxHeartbeats 600000 -- next def times out with default heartbeats (200000)
```

Comment:
That's unfortunate. Can we find someone to debug? Does it help to squeeze the simps?

Reply:
See: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/requesting.20help.20on.20mathlib4.23674

Reply:
I identified the slow simp and discovered that `rfl` closes the goal quickly.